agent/mapcall/src/main/java/randoop/instrument/ReplacementFileReader.java

```diff
@@ -0,0 +1,524 @@
+   */
+  private static void findPackageReplacements(
+      ConcurrentHashMap<MethodDef, MethodDef> replacementMap,
+      String original,
+      String replacement,
+      ClassLoader loader)
+      throws ReplacementFileException {
+    Enumeration<URL> resources;
+    try {
+      resources = loader.getResources(replacement.replace('.', '/'));
+    } catch (IOException e) {
+      throw new ReplacementFileException(e.getMessage());
+    }
+    while (resources.hasMoreElements()) {
+      URL url = resources.nextElement();
+      try {
+        URLConnection connection = url.openConnection();
+        if (connection instanceof JarURLConnection) {
+          JarFile jarFile = ((JarURLConnection) connection).getJarFile();
+          addPackageReplacements(replacementMap, original, replacement, jarFile);
+        } else {
+          // The subclass for directories is is internal.  It seems to work to assume the
+          // connection is a directory, and let an exception occur if it is not
+          File path = new File(URLDecoder.decode(url.getPath(), "UTF-8"));
+          if (path.exists() && path.isDirectory()) {
+            addPackageReplacements(replacementMap, original, replacement, path);
+            return;
+          }
+        }
+      } catch (IOException e) {
+        String msg =
+            String.format(
+                "Error identifying replacement %s with a package: %s", replacement, e.getMessage());
+        throw new ReplacementFileException(msg);
+      }
+    }
+    String msg =
+        String.format("No package or class for replacement %s found on classpath", replacement);
+    throw new ReplacementFileException(msg);
+  }
+
+  /**
+   * Discovers and adds method replacements for the package replacement determined by the {@code
+   * original} and {@code replacement} strings found in the given {@code ClassLoader}.
+   *
+   * @param replacementMap the method replacement map to which new replacements are added
+   * @param original the original package name
+   * @param replacement the replacement package name
+   * @param jarFile the jar file to search
+   * @return true if a class in the replacement package is found in the jar, false otherwise
+   */
+  private static boolean addPackageReplacements(
+      ConcurrentHashMap<MethodDef, MethodDef> replacementMap,
+      String original,
+      String replacement,
+      JarFile jarFile)
+      throws ReplacementFileException {
+
+    boolean found = false;
+    Enumeration<JarEntry> entries = jarFile.entries();
+    while (entries.hasMoreElements()) {
+      JarEntry entry = entries.nextElement();
+      String filename = entry.getName();
+      if (filename.endsWith(".class") && filename.startsWith(replacement.replace('.', '/'))) {
+        final String classname =
+            filename.substring(replacement.length() + 1, filename.lastIndexOf(".class"));
+        final String originalClassname = original + "." + classname;
```

Comment:
It is a but surprising to have a local variable for `originalClassname` but not for `replacementClassname` which is computed where it is needed.  This took me extra time to understand and verify the correctness of.  I suggest having a variable for both of them, for parallelism and consistency and to make the code easier to read and understand.  There is no negative performance impact because the JIT will only compute `replacementClassname` when needed.

There are some other places where this same comment applies.